oasis-reference/src/main/kotlin/com/empowerops/volition/ref_oasis/OptimizerEndPoint.kt

```diff
@@ -0,0 +1,271 @@
+            modelService.renameSim(target, newName, oldName)
+            NodeNameChangeResponseDTO.newBuilder().setChanged(true).build()
+        } else {
+            NodeNameChangeResponseDTO.newBuilder().setChanged(false).build()
+        }
+    }
+
+    override fun register(request: RegistrationCommandDTO, responseObserver: StreamObserver<OASISQueryDTO>) {
+        if (request.name in modelService.simByName.keys) {
+            return
+        }
+        modelService.addNewSim(request.name!!, Simulation(emptyList(), emptyList(), "", responseObserver, Channel(1), Channel(1), Channel(1)))
+        modelService.updateStatusMessage("${request.name} registered")
+    }
+
+    override fun unregisterAll() {
+        throw NotImplementedException()
+    }
+
+    override fun startOptimization(request: StartOptimizationCommandDTO, responseObserver: StreamObserver<StartOptimizationResponseDTO>) = responseObserver.consume {
+        startOptimization(RandomNumberOptimizer())
+        StartOptimizationResponseDTO.newBuilder().setMessage("Started").setStarted(true).build()
+    }
+
+
+    override suspend fun startOptimization(optimizer: Optimizer) {
+        state = State.Running
+
+        for (simName in modelService.simByName.keys) {
+            syncConfigFor(simName)
+        }
+
+        modelService.updateStatusMessage("Evaluating...")
+
+        while (state == State.Running) {
+            var toolNumber = 1
+            for ((simName, sim) in modelService.simByName) {
+                modelService.updateStatusMessage("Evaluating: $simName ($toolNumber/${modelService.simByName.keys.size})")
+                val inputVector = optimizer.generateInputs(sim.inputs)
+                val message = OASISQueryDTO.newBuilder()
+                        .setEvaluationRequest(OASISQueryDTO.SimulationEvaluationRequest.newBuilder()
+                                .setName(simName)
+                                .putAllInputVector(inputVector)
+                        )
+                        .build()
+
+                sim.input.onNext(message)
+                val result = select<SimResult> {
+                    sim.output.onReceive { SimResult.Success(it.name, it.outputVectorMap) }
+                    sim.error.onReceive { SimResult.Failure(it.name, it.exception) }
+                    if (sim.timeOut != null) {
+                        onTimeout(sim.timeOut.toMillis()) {
+                            //Timed out
+                            val message = OASISQueryDTO.newBuilder().setCancelRequest(
+                                    OASISQueryDTO.SimulationCancelRequest.newBuilder().setName(simName)
+                            ).build()
+                            sim.input.onNext(message)
+                            select {
+                                sim.output.onReceive { SimResult.TimeOut(it.name) }
+                                sim.error.onReceive { SimResult.TimeOutFailure(it.name, it.exception) }
```

Comment:
interestingly the timeout doesnt have a timeout, and it would be very difficult to give it one I suppose.

Reply:
Looking back at this, maybe every call should have a time out, I think I will talk about that later.